spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+On the other hand, methods for getting data from the [=shared storage database=] are declared in {{WorkletSharedStorage}} and exposed to the {{SharedStorageWorklet}} only, in order to carefully control the flow of data read from the [=shared storage database|database=].
+
+<xmp class='idl'>
+  [Exposed=(Window,SharedStorageWorklet)]
+  interface SharedStorage {
+    Promise<any> set(DOMString key,
+                     DOMString value,
+                     optional SharedStorageSetMethodOptions options = {});
+    Promise<any> append(DOMString key,
+                        DOMString value);
+    Promise<any> delete(DOMString key);
+    Promise<any> clear();
+  };
+
+  dictionary SharedStorageSetMethodOptions {
+    boolean ignoreIfPresent = false;
+  };
+</xmp>
+
+  ## The {{WindowSharedStorage}} interface ## {#window}
+
+  The {{WindowSharedStorage}} interface is as follows.
+
+<xmp class='idl'>
+  [Exposed=(Window)]
+  interface WindowSharedStorage : SharedStorage {
+    Promise<any> run(DOMString name,
+                     optional SharedStorageRunOperationMethodOptions options = {});
+    Promise<FencedFrameConfig> selectURL(DOMString name,
                                  FrozenArray<SharedStorageUrlWithMetadata> urls,
                                  optional SharedStorageRunOperationMethodOptions options = {});
 
-  Promise<any> run(DOMString name, optional SharedStorageRunOperationMethodOptions options = {});
+    readonly attribute SharedStorageWorklet worklet;
+  };
+
+  dictionary SharedStorageRunOperationMethodOptions {
+    object data;
+  };
+</xmp>
+
+  ### Window Setter/Deleter Methods ### {#window-setter}
+
+  <div algorithm>
+    The <dfn method for="WindowSharedStorage">set(|key|, |value|, |options|)</dfn> method steps are:
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |context| be {{WindowSharedStorage}}'s {{Window}}'s [=Window/browsing context=].
+    1. [=Assert=] that |context| is not null.
+    1. If |key|'s [=string/length=] exceeds the [=key/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |value|'s [=string/length=] exceeds the [=value/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |options| is invalid, reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |options| is empty, create |options|.
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |result| be null.
+        1. If |options|["`ignoreIfPresent`"] is true:
+            1. Let |currentValue| be the result of running [=shared storage database/retrieve an entry from the database=] with |context| and |key|.
+            1. If |currentValue| is not null and |currentValue| is not undefined, resolve |promise| as undefined.
+            1. If |currentValue| is null, reject |promise|.
```

Comment:
What is the significance of the current value being `null`?  Why does this cause the operation to fail instead of setting the new value?

Reply:
Hmm, many specs use verbiage like "return failure" and then "if |result| is failure" to signify this sort of thing.  That might be more clear.